### PR TITLE
Handle TiDB write-write conflict when optimistic transaction mo…

### PR DESCRIPTION
**Pull request by kalfonso**
…de is on.

When optimistic transaction mode is on TiDB detects write-write conflicts on commit. These exceptions must be mapped to Hibernate and retried.
https://docs.pingcap.com/tidb/stable/optimistic-transaction
https://docs.pingcap.com/tidb/stable/optimistic-transaction

### CODE REVIEW

**Comment by tirsen** [comment]
I really like this!

File: misk-hibernate/src/test/kotlin/misk/hibernate/TransacterTest.kt [hunk 1]
```diff
@@ -760,6 +763,39 @@ abstract class TransacterTest {
         "retried Movies transaction succeeded \\(attempt 3, same connection\\)")
   }
 
+  @Test
+  fun concurrentUpdates() {
```
Comment (zcross) [inline misk-hibernate/src/test/kotlin/misk/hibernate/TransacterTest.kt:767]: Great test 💯

**Review by zcross — APPROVED** [review]
I realize this is marked WIP and my review is unsolicited, but LGTM if you are trying to get this in.

File: misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt [hunk 2]
```diff
@@ -258,6 +254,23 @@ internal class RealTransacter private constructor(
     }
   }
 
+  private fun mapTidbException(e: Throwable): Throwable {
```
Comment (zcross) [inline misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt:257]: Superficial feedback: you could make this an extension: 

```
private fun Throwable.mapTidbException(): Throwable = when {
 ...
}
```

Then line 239 becomes
```
if (config.type == DataSourceType.TIDB) {
  rethrow = e.mapTidbException()
}
```

File: misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt [hunk 3]
```diff
@@ -391,6 +405,15 @@ internal class RealTransacter private constructor(
         message.contains("restart transaction")
   }
 
+  /**
+   * "Transactions in TiKV encounter write conflicts". This can happen when optimistic transaction
+   * mode is on. Conflicts are detected during transaction commit
+   * https://docs.pingcap.com/tidb/dev/tidb-faq#error-9007-hy000-write-conflict
+   */
+  private fun isTidbWriteConflict(th: SQLException): Boolean {
```
Comment (zcross) [inline misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt:413]: Does checking the mode when encountering this exception matter? Or is it only possible in optimistic mode?

File: misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt (hunk 2 above)
Comment (mightyguava) [reply misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt:257]: Making it an extension doesn't seem to read any better?

File: misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt (hunk 3 above)
Comment (mightyguava) [reply misk-hibernate/src/main/kotlin/misk/hibernate/RealTransacter.kt:413]: This is only possible in optimistic mode. In pessimistic mode the first write would acquire a lock and the second write would have to wait.

**Review by mightyguava — APPROVED** [review]
I updated the comments a bit, LGTM

**Comment by mightyguava** [comment]
Failed unrelated SQS tests, so just gonna merge

**Pull request merged**